README.md
# X32 Mute Matrix

[![Greenkeeper badge](https://badges.greenkeeper.io/lange/x32-mute-matrix.svg)](https://greenkeeper.io/)
[![Windows status](https://ci.appveyor.com/api/projects/status/rgtswyiipcqe4t4p?svg=true)](https://ci.appveyor.com/project/Lange/x32-mute-matrix)
[![Linux & OSX status](https://travis-ci.org/Lange/x32-mute-matrix.svg?branch=master)](https://travis-ci.org/Lange/x32-mute-matrix)

> See and configure the mute status of all your Behringer X32's channels (on every mixbus) in one place.

![screenshot](screenshot.png)

## Motivation
Sometimes, as the operator of an audio mixer, you just really want to see (and configure) the mute status of every channel on every bus. This program enables you to do this for your Behringer X32/M32 digital mixers.

## Features
- Manage and monitor the mute status of all 32 main inputs and 8 aux inputs on the Main bus, all 16 mixbuses, and the mono bus.
- Automatically pulls the names and colors of every bus/channel from the X32.
- Connection history that remembers device addresses that have been connected to in the past.
- Connection status indicator
- Automatic reconnection to previous device on program start
- Installers available for Windows, macOS, and Linux.

## Planned Features
- Stereo bus on/off
- Automatic discovery of X32/M32 devices
- Collapsing/folding of linked channels & buses
- Choose which channels and buses to show or hide
- Set the names and colors of channels and buses
- iOS and Android support

## Installation
Check the [Releases](https://github.com/lange/x32-mute-matrix/releases) page to grab the latest installer for your operating system.
Once installed, the application will autoupdate.

## Credits
Developed by [Alex Van Camp](https://twitter.com/vancamp)  
Assets and additional design by [Chris Hanel](https://twitter.com/chrishanel)  
QA and feedback provided by [Mark Schwartzkopf](https://twitter.com/m_schwartzkopf)

## License
X32 Mute Matrix is provided under the MIT license, which is available to read in the 
[LICENSE](https://github.com/lange/x32-mute-matrix/blob/master/LICENSE) file.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.1"></a>
## [1.0.1](https://github.com/lange/x32-mute-matrix/compare/v1.0.0...v1.0.1) (2017-11-09)


### Bug Fixes

* **autoupdater:** fix autoupdater never checking for updates ([a494e45](https://github.com/lange/x32-mute-matrix/commit/a494e45))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/lange/x32-mute-matrix/compare/v0.0.3...v1.0.0) (2017-11-09)


### Bug Fixes

* fix highlight crosshair being lost sometimes when toggling a channel's mute status ([ba5bc92](https://github.com/lange/x32-mute-matrix/commit/ba5bc92))
* ignore traffic from unexpected sources ([bcbdd16](https://github.com/lange/x32-mute-matrix/commit/bcbdd16))
* remove leading zeroes from row bus numbers ([68df684](https://github.com/lange/x32-mute-matrix/commit/68df684))
* use more appropriate cursors on channel buttons and labels ([8b4cb87](https://github.com/lange/x32-mute-matrix/commit/8b4cb87))


### Features

* add connection status indicator; redesign connection dialog ([3dc99a9](https://github.com/lange/x32-mute-matrix/commit/3dc99a9))
* add support for click-and-drag toggling of mutes ([57f514f](https://github.com/lange/x32-mute-matrix/commit/57f514f)), closes [#1](https://github.com/lange/x32-mute-matrix/issues/1)
* make channel labels sit at an angle for improved legibility ([8169bdf](https://github.com/lange/x32-mute-matrix/commit/8169bdf))
* redesign labels, differentiate between inverted and non-inverted colors ([a59d972](https://github.com/lange/x32-mute-matrix/commit/a59d972))


### Performance Improvements

* **client:** greatly improve general UI performance & framerate ([f951033](https://github.com/lange/x32-mute-matrix/commit/f951033))
* **x32-row:** use window.x32AppDarkenedOpacity for setting label opacity ([39d4404](https://github.com/lange/x32-mute-matrix/commit/39d4404))



<a name="0.0.3"></a>
## [0.0.3](https://github.com/lange/x32-mute-matrix/compare/v0.0.2...v0.0.3) (2017-09-07)


### Features

* implement UI for accepting automatic updates ([c22b015](https://github.com/lange/x32-mute-matrix/commit/c22b015))
* slightly reduce crosshair dimming ([c5007d4](https://github.com/lange/x32-mute-matrix/commit/c5007d4))



<a name="0.0.2"></a>
## [0.0.2](https://github.com/lange/x32-mute-matrix/compare/v0.0.1...v0.0.2) (2017-09-05)


### Bug Fixes

* **about:** use correct LICENSE and repo urls ([6a4a845](https://github.com/lange/x32-mute-matrix/commit/6a4a845))
* **build:** change how and when artifacts are published ([4dbe910](https://github.com/lange/x32-mute-matrix/commit/4dbe910))


### Features

* add support for Aux Ins ([4109016](https://github.com/lange/x32-mute-matrix/commit/4109016))
* add support for Main and Mono buses ([afadac2](https://github.com/lange/x32-mute-matrix/commit/afadac2))


### Performance Improvements

* simplify mix mutes fetching code, send less updates to client ([3274eef](https://github.com/lange/x32-mute-matrix/commit/3274eef))


app/client/elements/x32-app.js
(function () {
	'use strict';

	const {ipcRenderer} = require('electron');
	Polymer.setPassiveTouchGestures(true); // Added in Polymer v2.1.0

	/**
	 * @customElement
	 * @polymer
	 */
	class X32App extends Polymer.Element {
		static get is() {
			return 'x32-app';
		}

		static get properties() {
			return {
				buses: {
					type: Array,
					value() {
						const arr = [{
							name: 'main',
							label: 'Main'
						}];

						for (let i = 0; i < 16; i++) {
							arr.push({
								name: `mixbus${toFixed2(i)}`
							});
						}

						arr.push({
							name: 'mono',
							label: 'Mono'
						});

						return arr;
					}
				},
				channels: {
					type: Array,
					value() {
						const arr = [];
						for (let i = 0; i < 40; i++) {
							arr.push({});
						}
						return arr;
					}
				},
				highlightRow: {
					type: Number,
					reflectToAttribute: true,
					value: null
				},
				highlightColumn: {
					type: Number,
					reflectToAttribute: true,
					value: null
				}
			};
		}

		ready() {
			this._updateHighlights = this._updateHighlights.bind(this);
			super.ready();

			this.buses.forEach((bus, index) => {
				this.set(`buses.${index}.channels`, clone(this.channels));
			});

			ipcRenderer.on('x32-mutes', (event, mutes, mutesKeyOrder) => {
				this.buses.forEach((bus, busIndex) => {
					bus.channels.forEach((channel, channelIndex) => {
						const values = mutes[mutesKeyOrder[busIndex]];
						this.set(`buses.${busIndex}.channels.${channelIndex}.muted`, !values[channelIndex]);
					});
				});
			});

			ipcRenderer.on('x32-configs', (event, configs) => {
				for (const key in configs) {
					if (!{}.hasOwnProperty.call(configs, key)) {
						continue;
					}

					const index = parseInt(key.match(/\d+$/)[0], 10);
					let path = '';
					switch (true) {
						case key.startsWith('channel'):
							path = `channels.${index - 1}`;
							break;
						case key.startsWith('auxin'):
							path = `channels.${index + 31}`;
							break;
						case key.startsWith('mixbus'):
							path = `buses.${index}`;
							break;
						default:
							// Do nothing.
					}

					this.set(`${path}.label`, configs[key].label);
					this.set(`${path}.color`, configs[key].color);
				}
			});

			ipcRenderer.on('updateDownloaded', (event, info) => {
				this.$['updateDialog-label'].innerText = `A new version (${info.version}) is ready to install. Would you like to install it now?`;
				this.$.updateDialog.open();
			});

			ipcRenderer.send('init');
		}

		connectedCallback() {
			super.connectedCallback();
			Polymer.RenderStatus.beforeNextRender(this, () => {
				// This is a performance optimization, which prevents
				// needing to continually re-evaluate the CSS Custom Property.
				window.x32AppDarkenedOpacity = parseFloat(this.readCSSCustomProperty('--x32-app-dim-opacity'));
			});
		}

		readCSSCustomProperty(prop) {
			if ('ShadyCSS' in window) {
				return window.ShadyCSS.getComputedStyleValue(this, prop);
			}

			return getComputedStyle(this, prop);
		}

		_debounceUpdateHighlights() {
			this._debouncedDangThing = Polymer.Debouncer.debounce(
				this._debouncedDangThing,
				Polymer.Async.timeOut.after(0),
				this._updateHighlights
			);
		}

		_updateHighlights() {
			// Sometimes, toggling a channel's mute status seems to cause the
			// `_handleRowMouseLeave` and `_handleColumnMouseLeave` events to fire,
			// even if the mouse hasn't moved at all. I couldn't figure out why this was happening,
			// but I was able to implement this weird hack which double-checks what element is
			// actually under the cursor. If the pending highlights don't agree with what is actually
			// under the cursor, then we "recover" what the highlight actually should be based on
			// what we found under the cursor.
			// This is pretty ugly, but it works so whatever.
			// - Lange, 2017/11/07
			if (typeof this._lastMouseScreenX === 'number' &&
				typeof this._lastMouseScreenY === 'number' &&
				this._pendingHighlightColumn === null &&
				this._pendingHighlightRow === null) {
				const elem = this.shadowRoot.elementFromPoint(this._lastMouseScreenX, this._lastMouseScreenY);
				if (elem) {
					if (elem.tagName.toLowerCase() === 'x32-row') {
						this._pendingHighlightRow = elem.index;
						const elemInRow = elem.shadowRoot.elementFromPoint(this._lastMouseScreenX, this._lastMouseScreenY);
						if (elemInRow && elemInRow.tagName.toLowerCase() === 'x32-channel') {
							this._pendingHighlightColumn = elemInRow.index;
						}
					} else if (elem.tagName.toLowerCase() === 'x32-channel-label') {
						this._pendingHighlightColumn = elem.index;
					}
				}
			}

			this.highlightColumn = this._pendingHighlightColumn;
			this.highlightRow = this._pendingHighlightRow;
		}

		_handleColumnLabelMouseEnter(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightColumn = e.target.index;
			this._pendingHighlightRow = 'all';
			this._debounceUpdateHighlights();
		}

		_handleColumnLabelMouseLeave(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightColumn = null;
			this._pendingHighlightRow = null;
			this._debounceUpdateHighlights();
		}

		_handleRowMouseEnter(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightRow = e.target.index;
			this._debounceUpdateHighlights();
		}

		_handleRowMouseLeave(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightRow = null;
			this._debounceUpdateHighlights();
		}

		_handleColumnMouseEnter(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightColumn = e.detail.column;
			this._debounceUpdateHighlights();
		}

		_handleColumnMouseLeave(e) {
			this._lastMouseScreenX = e.x;
			this._lastMouseScreenY = e.y;
			this._pendingHighlightColumn = null;
			this._debounceUpdateHighlights();
		}

		_handleUpdateDialogClosed(e) {
			if (e.detail.confirmed) {
				ipcRenderer.send('installUpdateNow');
			}
		}
	}

	function clone(object) {
		return JSON.parse(JSON.stringify(object));
	}

	function toFixed2(num) {
		num += 1;
		return num < 10 ? `0${num}` : num;
	}

	customElements.define(X32App.is, X32App);
})();
